src/scf/newton.jl

```diff
@@ -0,0 +1,360 @@
+"""
+function NewtonConvergenceDensity(tolerance)
+    info -> (norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol) < tolerance)
+end
+
+"""
+Default callback function for `newton`, which prints a convergence table
+"""
+function NewtonDefaultCallback()
+    prev_energies = nothing
+    function callback(info)
+        !mpi_master() && return info  # Printing only on master
+        if info.stage == :finalize
+            info.converged || @warn "Newton not converged."
+            return info
+        end
+        collinear = info.basis.model.spin_polarization == :collinear
+
+        if info.n_iter == 1
+            E_label = haskey(info.energies, "Entropy") ? "Free energy" : "Energy"
+            magn    = collinear ? ("   Magnet", "   ------") : ("", "")
+            @printf "n     %-12s      Eₙ-Eₙ₋₁     ρ_next-ρ%s\n" E_label magn[1]
+            @printf "---   ---------------   ---------   --------%s\n" magn[2]
+        end
+        E    = isnothing(info.energies) ? Inf : info.energies.total
+        Δρ   = norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol)
+        if size(info.ρ_next, 4) == 1
+            magn = NaN
+        else
+            magn = sum(spin_density(info.ρ_next)) * info.basis.dvol
+        end
+
+        Estr   = (@sprintf "%+15.12f" round(E, sigdigits=13))[1:15]
+        prev_E = prev_energies === nothing ? Inf : prev_energies.total
+        ΔE     = prev_E == Inf ? "      NaN" : @sprintf "% 3.2e" E - prev_E
+        Mstr = collinear ? "   $((@sprintf "%6.3f" round(magn, sigdigits=4))[1:6])" : ""
+        @printf "% 3d   %s   %s   %2.2e%s\n" info.n_iter Estr ΔE Δρ Mstr
+        prev_energies = info.energies
+
+        flush(stdout)
+        info
+    end
+end
+
+# Newton algorithm
+
+"""
+    newton_step(basis::PlaneWaveBasis, ψ, res, occupation;
+                tol_cg=1e-10, verbose=false)
+
+Perform a Newton step : we take as given a planewave set ψ and we return the
+Newton step δψ and the updated state ψ - δψ
+(after proper orthonormalization) where δψ solves Jac * δψ = res
+and Jac is the Jacobian of the projected gradient descent : Jac = Ω+K.
+δψ is an element of the tangent space at ψ (set to ψ if not specified in newton function).
+"""
+function newton_step(basis::PlaneWaveBasis, ψ, res, occupation;
+                     tol_cg=1e-10, verbose=false)
+
+    N = size(ψ[1], 2)
+    Nk = length(basis.kpoints)
+
+    # compute quantites at the point which define the tangent space
+    ρ = compute_density(basis, ψ, occupation)
+    energies, H = energy_hamiltonian(basis, ψ, occupation; ρ=ρ)
+
+    # packing routines
+    # some care is needed here : K is real-linear but not complex-linear because
+    # of the computation of δρ from δψ. To overcome this difficulty, instead of
+    # seeing the jacobian as an operator from C^Nb to C^Nb, we see it as an
+    # operator from R^2Nb to R^2Nb. In practice, this is done with the
+    # reinterpret function from julia. Thus, we are sure that the map we define
```

Comment:
Maybe move functions real_to_complex and complex_to_real in packing?

Reply:
I'm not sure, it is not necessary for direct_minimization.

Reply:
yeah but they are generally useful operations related to packing, so might as well put them there

Reply:
Ok, but how do we do with `direct_minimization.jl` then ? This R2I and I2R stuff is not needed right ?

Reply:
That's actually interesting. The LBFGS implementation in optim was just converted from real to complex blindly (I did it in a hurry a few years ago). So it's basically approximating the Hessian of f as a complex matrix, which is only valid is the gradient is holomorphic (which it's not in our case). So it would probably be better to do direct_minimization in real arithmetic as well! Especially in models like gross-pitaevskii where the nonlinearity is important. Could you try it eg on gross_pitaevskii_2D?